lib/puppet/provider.rb

```diff
@@ -384,7 +384,7 @@ def self.initvars
   # @raise [Puppet::DevError] Error indicating that the method should have been implemented by subclass.
   # @see prefetch
   def self.instances
-    raise Puppet::DevError, _("Provider %{provider} has not defined the 'instances' class method") % { provider: self.name }
+    raise Puppet::DevError, _("To support listing resources of this type the '%{provider}' provider needs to implement an 'instances' class method returning the current set of resources. See https://puppet.com/docs/puppet/latest/provider_development.html#prefetching for details or port your module to the Resource API for a much simpler experience: https://puppet.com/docs/puppet/6.4/create_types_and_providers_resource_api.html") % { provider: self.name }
```

Comment:
We should avoid subjective statements such as "for a much simpler experience" in any output. I'm also wary of adding links here that could change and aren't stable.

Reply:
We recently had a discussion about stable links with @jbondpdx, though I don't remember the result?

Reply:
1. I agree with @jtappa about "simpler". What is simple or easy for us may or may not be or the user, so we avoid these words in docs. Maybe use something like "...or port your module to the updated Resource API..." ?

2. Without a version, we don't have a good way to provide a stable link. Honestly, I have no problem with /latest/, assuming you actually want the most recent version and not specifically the doc associated with the relevant Puppet version. It's probably the least problematic way to provide an explicit link though. I'd suggest leaving the anchor off, because those are more subject to change than the base URL.

The trouble with versioned links is that they should be regularly updated in new Puppet versions. I like them for, say, deprecations or version specific release notes, but they are a maintenance chore in more general contexts. 

It might be better to avoid explicit links and just use searchable language like "Refer to our provider development docs at puppet.com/docs".

Reply:
* updated the message to link directly to a search that (as of today) links the reader to the right page
* removed the link to the provider page, which doesn't have an example anyways
* moved the emphasis of "simpler" from "experience" to "Resource API". The Resource API is objectively simpler than the type and provider "DSL" in puppet core. Porting over providers usually reduces LoC by 30% and allows(!) achieving 100% unit test coverage, which was previously impossible.

Reply:
I'll give up my objection to "simple" if you please please remove "much", @DavidS . :)

Reply:
deal and done!